[MRG] ChainedImputer -> IterativeImputer, and documentation update

sergeyf opened the pull request:
Addresses two points as discussed in #11259:
(a) Removes the "average last `n_imputations` behavior". Now there just one parameter `n_iter` instead of the two: `n_burn_in` and `n_imputations`.
(b) New flag: `sample_after_predict`, which is `False` by default. If true it will sample from the predictive posterior after predicting during each round-robin iteration. Turning it on will make `ChainedImputer` run a single 

@jnothman @glemaitre: I think `ChainedImputer` needs a new check: if `sample_after_predict=True`, then the predictor needs to have `return_std` as a parameter of its `predict` method. What's best practice for checking something like that?

Also, we may want a new example of how to use `ChainedImputer` for the purpose of a MICE-type analysis, and how to use `ChainedImputer` with a `RandomForest` as the predictor instead of `BayesianRidge` to demonstrate missForest functionality. 

I think @RianneSchouten would be a good candidate for the MICE example as a contribution to this PR or a new one, and I can stick `ChainedImputer` + RF into `plot_missing_values.py` as another bar on the plot. Let me know how that sounds.

sergeyf committed:
first commit

sergeyf committed:
complete functional overhaul

sergeyf committed:
fixing some documentation

jnothman commented:
CIs failing, FYI

sergeyf committed:
fixing broken test

sergeyf commented:
Fixed, thank you!

RianneSchouten commented:
I will work on the example next wednesday and make a PR then.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virusvrij.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

2018-06-25 7:44 GMT+02:00 Sergey Feldman <notifications@github.com>:

> Fixed, thank you!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11350#issuecomment-399837999>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVQqew6SZKjfI7oEWczNqwS6XHmkYw5zks5uAHihgaJpZM4U1b8P>
> .
>

jnothman commented on sklearn/impute.py:
```diff
@@ -452,17 +452,19 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        It must support ``return_std`` in its ``predict`` method if
+        ``sample_after_predict`` option is set to ``True`` below.
+
+    sample_after_predict : boolean, default=False
```
I think this name is unnecessarily verbose.

Brainstorming.

* fill_with='expectation' vs 'sample'
* random_impute bool
* use_std
* sampled
* random_draw
* randomized (but there are other random components)

sergeyf replied:
How about `predict_posterior` bool?

sergeyf committed:
addressing review comments + stef's suggestions for documentation

sergeyf commented:
@stefvanbuuren This PR has a lot of the documentation changes you suggested. Please take a look at what's here and let me know if we're addressing your concerns. What's left to do is to make an example that actually demonstrates how to use `ChainedImputer` as MICE.

sergeyf commented:
@jnothman I'd like to raise an error if a predictor without `return_std` is passed in when `predict_posterior=True`.  The only way I've found to do this is the `inspect` module as follows:

```
predictor = BayesianRidge()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # True

predictor = RandomForestRegressor()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # False
```

OK to rely on the `inspect` module in such a fashion in the middle of `impute.py`?

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -93,24 +93,26 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
-Then, the regressor is used to predict the unknown values of `y`. This is repeated
-for each feature in a chained fashion, and then is done for a number of imputation
-rounds. Here is an example snippet::
+A more sophisticated approach is to use the :class:`ChainedImputer` class, models
```
"model" -> "which models" ?

sergeyf replied:
Whoops, thanks.

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -93,24 +93,26 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
+estimate for imputation. It does so in an iterated round-robin fashion: at each step,
+a feature column is designated as output `y` and the other feature columns are treated
+as inputs `X`. A regressor is fit on `(X, y)` for known `y`. Then, the regressor is
+used to predict the unknown values of `y`. This is repeated for each feature in a
```
"unknown" -> "missing"? (only a suggestion, the unknown is fine as well, but the missing might be more clear since it about imputing the missing values)

sergeyf replied:
Changed, thanks.

jorisvandenbossche commented on sklearn/impute.py:
```diff
@@ -426,11 +426,11 @@ def transform(self, X):
-    of the features are Gaussian.
+    Basic implementation of chained mutual regressions to find replacement
+    values in multivariate missing data. This version assumes all features
+    are Gaussian.
```
Is it needed to mention this restriction of Gaussian? As I assume it depends on the predictor one uses.

sergeyf replied:
So far yes. The posterior sampling is done via a call to predict that also returns the standard deviation and then we sample from a Gaussian posterior. It's a limitation at the moment.

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -93,24 +93,26 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
+chained fashion, and then is done for a number of imputation rounds. The results
+of the final imputation round are returned. Our implementation was inspired by the
+R MICE package (Multivariate Imputation by Chained Equations), but differs from
+it in setting single imputation to default instead of multiple imputation. This
```
I find the "setting single imputation to default" still confusing. 
As AFAIK, the `ChainedImputer` itself can only do single imputation, that's not a matter of its default settings. You can of course then use it multiple times (setting the appropriate arguments for this case) to do multiple imputation.

stefvanbuuren replied:
Yes, you're right, but by now users have come to expect that "chained equations" refers to multiple imputation, so the just combination "chained + single" would confuse quite a few.

jorisvandenbossche reviewed (commented):
@sergeyf Thanks a lot for keeping up making changes based on the discussion! 
Added some quick comments on the docs.

stefvanbuuren commented:
@sergeyf I think it is clear now that your method does single imputation by default. 

Perhaps for balance also mention that what you describe as "the most common use case" differs from what is generally recommended by the statistical community (you can refer to Little & Rubin (2002, Ch 4) and Rubin (1987, Ch 1)), and only then say it's an open problem.

sergeyf committed:
fixing docs

sergeyf commented:
@stefvanbuuren this is the current full paragraph, which starts with the point about it being different in the statistics community: 

```
In the statistics community, it is common practice to perform multiple imputations,
generating, for example, 10 separate imputations for a single feature matrix.
Each of these 10 imputations is then put through the subsequent analysis pipeline
(e.g. feature engineering, clustering, regression, classification). The 10 final
analysis results (e.g. held-out validation error) allow the data scientist to
obtain understanding of the uncertainty inherent in the missing values. The above
practice is called multiple imputation. As implemented, the :class:`ChainedImputer`
class generates a single imputation for each missing value because this
is a common use case for machine learning applications. However, it can also
be used for multiple imputations by applying it repeatedly to the same dataset with
different random seeds.
```

I'll add a parenthetical reference to it.

sergeyf committed:
more update to impute.rst

jnothman commented:
I don't really like predict_posterior as the difference between the settings is not about whether prediction is from the posterior, but about whether we use the expectation or a random draw from that posterior

I don't think it is essential to raise an explicit error for the case that the estimator lacks return_std: the default TypeError will be informative enough when it tries to predict, and users who want to use the sampling variant will likely have read it's documentation which mentions return_std

sergeyf commented:
OK. thanks. I'll change the name to `posterior_draw`?

The only thing decision left is whether to change the name to `IterativeImputer`.

jnothman commented:
I'm +1 for IterativeImputer. But I was +1 for ChainedImputer too :) But I
suppose then it was more urgent to make it anything but MICE :P

On 26 June 2018 at 09:22, Sergey Feldman <notifications@github.com> wrote:

> OK. thanks. I'll change the name to posterior_draw?
>
> The only thing decision left is whether to change the name to
> IterativeImputer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11350#issuecomment-400126264>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wyVh0wTjWhKlGBhwQYw3g7bRw8Bks5uAXDHgaJpZM4U1b8P>
> .
>

sergeyf committed:
name change again, new default ridgeCV

sergeyf committed:
small bug

sergeyf committed:
output changed in impute.rst because of RidgeCV

sergeyf commented:
@glemaitre @jnothman I changed the default to be `RidgeCV` in case `sample_posterior=False`.  One thing that I saw is that some of the tests started failing. It looks like the default `alphas` in `RidgeCV` are only `[0.1, 1, 10]`. I'm worried that's too much regularization. So, I also added a small value in case no regularization is needed:

```
self._predictor = RidgeCV(alphas=np.array([1e-5, 0.1,  1, 10]))
```

Let me know if there are any other gotchas I should be aware of in instantiating a default `RidgeCV`.

Thanks!

jnothman commented on sklearn/impute.py:
```diff
@@ -836,13 +862,16 @@ def fit_transform(self, X, y=None):
-        # then there is no need to do burn in and the result should be
-        # just the initial imputation (before clipping)
-        if self.n_imputations < 1:
+        # edge case: in case the user specifies 0 for n_burn_in,
```
n_iter

jnothman commented on sklearn/impute.py:
```diff
@@ -908,21 +933,21 @@ def transform(self, X):
-        # then there is no need to do burn in and the result should be
-        # just the initial imputation (before clipping)
-        if self.n_imputations < 1:
+        # edge case: in case the user specifies 0 for n_burn_in,
```
n_iter

jnothman commented on sklearn/impute.py:
```diff
@@ -908,21 +933,21 @@ def transform(self, X):
             return X_filled
 
-        X_filled = np.clip(X_filled, self._min_value, self._max_value)
+        # clip only the initial filledin values
```
I don't really get why we would expect the initial imputation to require clipping.

sergeyf replied:
Huh, that's a good point. What I should probably do is this:

```
    if self.n_iter < 1:
        X_filled[mask_missing_values] = np.clip(X_filled[mask_missing_values],
                                                self._min_value,
                                                self._max_value)
        return X_filled
```

sergeyf replied:
Or actually just not clip before imputing.

jnothman commented on sklearn/impute.py:
```diff
@@ -935,15 +960,13 @@ def transform(self, X):
-                    Xt += X_filled
                 if self.verbose > 1:
-                    print('[ChainedImputer] Ending imputation round '
+                    print('[IterativeImputer] Ending imputation round '
```
This isn't covered in tests, FWIW. It should be covered to ensure there's no AttributeError or whatever

sergeyf replied:
Will do.

jnothman commented on sklearn/tests/test_impute.py:
```diff
@@ -486,44 +486,25 @@ def test_imputation_copy():
     # made, even if copy=False.
 
 
-def test_chained_imputer_rank_one():
```
This should still be valid if we run multiple imputations and take the mean, shouldn't it? Let's try leave the tests in.

sergeyf replied:
It still is: I changed the name to `test_iterative_imputer_rank_one` and moved it to be with the other `functional` tests.

jnothman commented on sklearn/tests/test_impute.py:
```diff
@@ -614,17 +592,17 @@ def test_chained_imputer_missing_at_transform(strategy):
                   initial_imputer.transform(X_test)[:, 0])
 
 
-def test_chained_imputer_transform_stochasticity():
```
This should be run with and without sample_posterior=True

sergeyf replied:
Will do.

jnothman reviewed (commented):
Please check that `sample_posterior=False` is deterministic if `n_nearest_features` is None.

I've also realised that `n_nearest_features` documentation fails to mention that the features are not necessarily the nearest, but are drawn with probability proportional to correlation.

(And now that we're supporting non-linear regresssors, I wonder if we should be using a rank correlation in `n_nearest_features`, or max(spearman_rho, pearson_rho) or something nasty like that. This `n_nearest_features` is useful but might get in the way of good results with non-linear prediction.)

Thanks for all this great work. I think when we're finished here, we'll have a very powerful and useful tool.

sergeyf commented on doc/modules/impute.rst:
```diff
@@ -133,12 +136,19 @@ Each of these 10 imputations is then put through the subsequent analysis pipelin
+As implemented, the :class:`IterativeImputer` class generates a single imputation
+for each missing value because this is the most common use case for machine learning
+applications. However, it can also be used for multiple imputations by applying it
+repeatedly to the same dataset with different random seeds.
```
Right, I should change this to say `set sample_posterior=True`.

sergeyf commented on doc/modules/impute.rst:
```diff
@@ -133,12 +136,19 @@ Each of these 10 imputations is then put through the subsequent analysis pipelin
+See Chapter 4 of "Statistical Analysis with Missing Data" by Little and Rubin for
+more discussion on multiple vs. single imputations.
+
+It is still an open problem as to how useful single vs. multiple imputation is in
```
Good addendum, thanks.

sergeyf commented on sklearn/impute.py:
```diff
@@ -452,17 +452,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
-    n_imputations : int, optional (default=100)
-        Number of chained imputation rounds to perform, the results of which
-        will be used in the final average.
+    n_iter : int, optional (default=10)
```
I'd prefer to leave it as-is until we actually support early stopping. The word `max` would be confusing as it is now.

sergeyf commented on sklearn/impute.py:
```diff
@@ -498,18 +505,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
     ----------
-    initial_imputer_ : object of class :class:`sklearn.preprocessing.Imputer`'
-        The imputer used to initialize the missing values.
+    initial_imputer_ : object of type :class:`sklearn.impute.SimpleImputer`
```
Another thing I'd like to defer to future Sergey =)

sergeyf committed:
Update impute.rst a smidge.

sergeyf committed:
Remove pre-clipping

sergeyf commented:
Thanks @jnothman - I'll get the last few to-dos done.

Re: `n_nearest_features`. This sounds like another PR, really. It would need a bit of investigation first.

sergeyf commented:
I just realized I don't know what this means: `Please check that sample_posterior=False is deterministic if n_nearest_features is None.` How are these two flags related? One is about sampling from the predictive posterior and one is about how many (nearest) features to use to fit each model.

jnothman commented:
I think all randomisation is due to n_nearest_features != None,
sample_posterior=True,
imputation_order='random', or predictor being randomised. Otherwise is
there any source of non-determinism??​

sergeyf commented:
Oh I see what you're saying. So I should check that, even if `random_state` is not fixed, but `sample_posterior=False`, `imputation_order=ascending`, `n_nearest_features=None`, then the results are still the same.

sergeyf commented:
Assuming `RidgeCV`'s cross-validation is also fixed?

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -87,37 +87,40 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
+for known `y`. Then, the regressor is used to predict the missing values of `y`.
+This is repeated for each feature in an iterative fashion, and then is done for a
+number of imputation rounds. The results of the final imputation round are
+returned. Our implementation was inspired by the R MICE package (Multivariate
```
Refer to a reference note to the paper of MICE.

sergeyf replied:
What's the standard way to do this? Just put in `[1]`?

glemaitre replied:
We can make a section reference in sphinx with a number and then refer to the number as well.

sergeyf replied:
Sorry, I'm not very experienced in sphinx. Is there an example somewhere else in sklearn's documentation that I can copy?

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -87,37 +87,40 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
+returned. Our implementation was inspired by the R MICE package (Multivariate
+Imputation by Chained Equations), but differs from it in setting single imputation
+to default instead of multiple imputation. This is discussed further below.
+Here is an example snippet::
```
Remove this sentence.

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -133,12 +136,21 @@ Each of these 10 imputations is then put through the subsequent analysis pipelin
 (e.g. feature engineering, clustering, regression, classification). The 10 final
```
I would remove the example of `10` and instead replace by `n` on the 2 lines above.

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -87,37 +87,40 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
-            n_imputations=10, n_nearest_features=None, predictor=None,
-            random_state=0, verbose=False)
+    IterativeImputer(imputation_order='ascending', initial_strategy='mean',
+             max_value=None, min_value=None, missing_values=nan, n_iter=10,
```
I would add a small description on the parameter which are important: `n_iter` might be probably enough.

sergeyf replied:
I added in above paragraph.

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -133,12 +136,21 @@ Each of these 10 imputations is then put through the subsequent analysis pipelin
+repeatedly to the same dataset with different random seeds when
+``sample_posterior=True``.
+
+See Chapter 4 of "Statistical Analysis with Missing Data" by Little and Rubin for
```
Put a reference section and refer to it..

glemaitre commented on examples/plot_missing_values.py:
```diff
@@ -8,7 +8,7 @@
 which could dominate results (otherwise known as a 'long tail').
 
-Another option is the ``ChainedImputer``. This uses round-robin linear
+Another option is the ``IterativeImputer``. This uses round-robin linear
```
I think that we can use the sphinx :class: directive

glemaitre commented on sklearn/impute.py:
```diff
@@ -452,17 +452,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+    predictor : estimator object, default=BayesianRidge() or RidgeCV()
         The predictor to use at each step of the round-robin imputation.
-        It must support ``return_std`` in its ``predict`` method.
+        It must support ``return_std`` in its ``predict`` method if
```
I would inverse the sentence.

```
If ``sample_posterior`` is True, the predictor, ....
```

glemaitre commented on sklearn/impute.py:
```diff
@@ -452,17 +452,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        fitted predictor for each imputation. Predictor must support
+        ``return_std`` in its ``predict`` method if set to ``True``. Set to
+        ``True`` if using ``IterativeImputer`` to have the same functionality
+        as MICE.
```
Be careful. MICE is not defined.

glemaitre commented on sklearn/impute.py:
```diff
@@ -498,18 +505,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+    imputation is supported with multiple instances of the imputer with
+    different random seeds run in parallel.
+
     The R version of MICE does not have inductive functionality, i.e. first
```
I am not sure to mention this anymore since that we are doing which is not MICE.

sergeyf replied:
Good point.

glemaitre commented on sklearn/impute.py:
```diff
@@ -498,18 +505,24 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
 
     Notes
     -----
+    This implementation was inspired by the R MICE package (Multivariate
```
Do we want only to mention about the R MICE package in the user guide instead of the docstring.

sergeyf replied:
I personally think it's OK here also.

glemaitre commented on sklearn/impute.py:
```diff
@@ -624,12 +642,15 @@ def _impute_one_feature(self,
-        imputed_values[~good_sigmas] = mus[~good_sigmas]
-        imputed_values[good_sigmas] = self.random_state_.normal(
-            loc=mus[good_sigmas], scale=sigmas[good_sigmas])
+        if self.sample_posterior:
```
I would make a check that the signature of `predictor.predict` has a `return_std` to raise a proper error.

sergeyf replied:
@jnothman suggested it's not necessary above. The only way I found to do this is:

```
predictor = BayesianRidge()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # True

predictor = RandomForestRegressor()
'return_std' in list(inspect.signature(predictor.predict).parameters.keys()) # False
```

glemaitre replied:
Uhm that's ugly.

In the same time, it is true that the error is ok
```
In [8]: reg.predict(X, predict_std=True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-d45a86188634> in <module>()
----> 1 reg.predict(X, predict_std=True)

TypeError: predict() got an unexpected keyword argument 'predict_std'
```

sergeyf replied:
Agreed on both points. Let's leave as is then.

jnothman replied:
Fwiw you don't need list or keys in the snippet above

glemaitre commented on sklearn/impute.py:
```diff
@@ -825,8 +846,13 @@ def fit_transform(self, X, y=None):
+            else:
+                from .linear_model import RidgeCV
+                # including a very small alpha to approximate OLS
+                self._predictor = RidgeCV(alphas=np.array([1e-5, 0.1,  1, 10]))
```
You will need to set the random state of the estimator.

glemaitre replied:
if `random_state is not None`

sergeyf replied:
I'm confused, there is no `random_state` anywhere here: http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html

glemaitre replied:
Uhm. I have to check the code. I could have think that the CV was actually relying on that but I might be wrong.

sergeyf replied:
I looked the `RidgeCV` code - no trace of random state!

glemaitre commented on sklearn/impute.py:
```diff
@@ -825,8 +846,13 @@ def fit_transform(self, X, y=None):
+                # including a very small alpha to approximate OLS
+                self._predictor = RidgeCV(alphas=np.array([1e-5, 0.1,  1, 10]))
         else:
             self._predictor = clone(self.predictor)
```
Actually we should set the state here also if `_predictor` has a `random_state`.

sergeyf replied:
Right, good point.

sergeyf replied:
Actually, shouldn't that be done on line 857 instead? I.e. just once?
```
        if self.predictor is None:
            if self.sample_posterior:
                from .linear_model import BayesianRidge
                self._predictor = BayesianRidge()
            else:
                from .linear_model import RidgeCV
                # including a very small alpha to approximate OLS
                self._predictor = RidgeCV(alphas=np.array([1e-5, 0.1,  1, 10]))
        else:
            self._predictor = clone(self.predictor)
        # SET RANDOM STATE HERE
```

glemaitre replied:
It is what I meant.

glemaitre commented on sklearn/impute.py:
```diff
@@ -877,14 +899,12 @@ def fit_transform(self, X, y=None):
+                      % (i_rnd + 1, self.n_iter, time() - start_t))
 
-        Xt /= self.n_imputations
+        Xt = X_filled
```
this could be simplified, isn't it?

glemaitre commented on sklearn/impute.py:
```diff
@@ -935,15 +950,13 @@ def transform(self, X):
                 i_rnd += 1
 
-        Xt /= self.n_imputations
+        Xt = X_filled
```
this could be simplified, isn't it?

sergeyf replied:
I'll replace `X_filled` with `Xt` everywhere.

glemaitre reviewed (changes requested):
Couple of suggestions.

sergeyf committed:
addressing comments

sergeyf committed:
forgotten doc change

sergeyf commented:
@jnothman I think I've addressed all your points.

@glemaitre I think I've addressed all your points, except that the references in `impute.rst` are formatted in a non-linking basic fashion instead of using sphinx.

jeremiedbb commented on sklearn/impute.py:
```diff
@@ -452,24 +452,34 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        as MICE (Multivariate Imputation by Chained Equations).
 
     n_nearest_features : int, optional (default=None)
         Number of other features to use to estimate the missing values of
```
just a typo here: 'of each' instead of 'of the each'

sergeyf replied:
Thanks!

jnothman commented:
RidgeCV is fully deterministic

sergeyf committed:
typo

RianneSchouten commented:
As promised, an example of using IterativeImputer as a MICE imputer in PR #11370 
Is someone available to review?

sergeyf commented:
@glemaitre Have I addressed all of your concerns?

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -87,37 +87,39 @@ string values or pandas categoricals when using the ``'most_frequent'`` or
+for known `y`. Then, the regressor is used to predict the missing values of `y`.
+This is done for each feature in an iterative fashion, and then is repeated for
+`n_iter` imputation rounds. The results of the final imputation round are
+returned. Our implementation was inspired by the R MICE package (Multivariate
```
It's important to give correct credit to MICE as an inspiration, but, purely text-wise, I think here it does not add much value, as the final result is really different from MICE. 
It would rather mention MICE below in the section of multiple imputation.

sergeyf replied:
Good idea.

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -128,17 +130,36 @@ Multiple vs. Single Imputation
+practice is called multiple imputation.
 
-Note that a call to the ``transform`` method of :class:`ChainedImputer` is not
+As implemented, the :class:`IterativeImputer` class generates a single imputation
```
I would keep this more general on imputers in sklearn, as also the SimpleImputer (which would be relevant in the case of a future 'sampling' strategy) is a single imputation. 
Then in the next sentence you can replace 'it' with 'the IterativeImputer'

sergeyf replied:
Will do, thanks.

jorisvandenbossche reviewed (commented):
Added some more comments on the documentation.

In addition, it would maybe be good to add a See Also reference to each other between SimpleImputer and IterativeImputer.

jorisvandenbossche commented on sklearn/impute.py:
```diff
@@ -423,14 +423,14 @@ def transform(self, X):
-class ChainedImputer(BaseEstimator, TransformerMixin):
-    """Chained imputer transformer to impute missing values.
+class IterativeImputer(BaseEstimator, TransformerMixin):
+    """Iterative imputer transformer to impute missing values.
```
Some brainstorming: use "Multivariate" to contrast it with the SimpleImputer? 

"Multivariate imputer that iteratively models each feature against the others" probably does not fit a single line ..

sergeyf replied:
Yeah, one line is tough =). I have this for now: `Multivariate imputer that estimates each features from all the others.`

stefvanbuuren commented:
@sergeyf Text is fine, but I would suggest saying: 
"... to obtain understanding of how analytic results may differ as a consequence of the inherent uncertainty caused by the missing values."

sergeyf commented on sklearn/impute.py:
```diff
@@ -498,23 +508,27 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
-    during the ``transform`` phase.
+    This implementation was inspired by the R MICE package (Multivariate
+    Imputation by Chained Equations), but differs from it in setting single
+    imputation to default instead of multiple imputation. However, multiple
```
Right, I missed that before. Fixed.

sergeyf commented on sklearn/impute.py:
```diff
@@ -452,24 +452,34 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        fitted predictor for each imputation. Predictor must support
+        ``return_std`` in its ``predict`` method if set to ``True``. Set to
+        ``True`` if using ``IterativeImputer`` to have the same functionality
+        as MICE (Multivariate Imputation by Chained Equations).
```
I've simplified.

sergeyf commented on sklearn/impute.py:
```diff
@@ -452,24 +452,34 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
-        will not be returned.
-
-    predictor : estimator object, default=BayesianRidge()
+    predictor : estimator object, default=BayesianRidge() or RidgeCV()
```
Check.

sergeyf commented on sklearn/impute.py:
```diff
@@ -423,14 +423,14 @@ def transform(self, X):
-    of the features are Gaussian.
+    Basic implementation of iterative mutual regressions to find replacement
+    values in multivariate missing data. This version assumes all features
+    are Gaussian.
```
Removed.

sergeyf commented on sklearn/impute.py:
```diff
@@ -423,14 +423,14 @@ def transform(self, X):
-    Imputations by Chained Equations) package from R. This version assumes all
-    of the features are Gaussian.
+    Basic implementation of iterative mutual regressions to find replacement
+    values in multivariate missing data. This version assumes all features
```
Replaced.

sergeyf committed:
addressing comments

sergeyf commented:
@stefvanbuuren @jorisvandenbossche Thanks for the suggestions. I've addressed most of them. 

Re: `n_iter` becoming `max_iter`. I can make the change, but let's make sure we agree that early stopping is something to definitely implement. It doesn't seem that important in my opinion, but I'm not too committed to that stance.

jnothman commented:
I don't think the idea of convergence makes sense when you're doing multiple imputation. I'm okay leaving it as `n_iter`.

stefvanbuuren commented:
For multiple imputation, we would like to have convergence in distribution. In typically applications of MICE convergence is very quick (say 5-10 iterations), since the inclusion of noise in the imputations makes the autocorrelations low, and hence convergence fast.

If you're imputing one single "best" value, then you'd want iterative convergence by optimizing some value. The behaviour of such an algorithm is quite different, and probably more like the EM-algorithm, which is known to be slow.

sergeyf commented:
Thanks @jnothman and @stefvanbuuren. Sounds like we can leave it as `n_iter`.

Is there anything else left to do on this PR? I don't have anything on my to-do list.

jnothman reviewed (commented):
Make the `random_state` docs more like:
```
	Randomizes selection of predictor features if n_nearest_features is not
	None, the imputation_order if 'random', and the sampling from posterior if
	sample_posterior is True. Use an integer for determinism. See :term:`the
	Glossary <random_state>`.
```

sergeyf committed:
fixing docs from review

sergeyf commented:
Thanks @jnothman. Fixed. 

@glemaitre I await your pleasure!

sergeyf committed:
Merge branch 'master' into chained_imputer_update

sergeyf commented:
@glemaitre Any chance for a final review?

glemaitre committed:
Merge remote-tracking branch 'origin/master' into chained_imputer_update

glemaitre committed:
TST update tests with renaming

glemaitre commented:
I am gonna make some fixes and we will merge this PR during the sprint

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -128,17 +128,37 @@ Multiple vs. Single Imputation
-random seeds with the ``n_imputations`` parameter set to 1.
-
-Note that a call to the ``transform`` method of :class:`ChainedImputer` is not
+generating, for example, `m` separate imputations for a single feature matrix.
```
double backsticks

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -128,17 +128,37 @@ Multiple vs. Single Imputation
-
-Note that a call to the ``transform`` method of :class:`ChainedImputer` is not
+generating, for example, `m` separate imputations for a single feature matrix.
+Each of these `m` imputations is then put through the subsequent analysis pipeline
```
double backsticks

glemaitre commented on doc/modules/impute.rst:
```diff
@@ -128,17 +128,37 @@ Multiple vs. Single Imputation
-Note that a call to the ``transform`` method of :class:`ChainedImputer` is not
+generating, for example, `m` separate imputations for a single feature matrix.
+Each of these `m` imputations is then put through the subsequent analysis pipeline
+(e.g. feature engineering, clustering, regression, classification). The `m` final
```
double backsticks

glemaitre commented on sklearn/impute.py:
```diff
@@ -461,24 +464,33 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        imputed target feature. Can provide significant speed-up when the
+        number of features is huge. If ``None``, all features will be used.
 
     initial_strategy : str, optional (default="mean")
```
We need to move this to accept any Imputer instead which can default to SimpleImputer('mean')

glemaitre commented on sklearn/impute.py:
```diff
@@ -835,66 +867,68 @@ def fit_transform(self, X, y=None):
+                from .linear_model import BayesianRidge
+                self._predictor = BayesianRidge()
+            else:
+                from .linear_model import RidgeCV
```
Let's default to RandomForestRegressor

glemaitre commented on sklearn/impute.py:
```diff
@@ -916,44 +950,36 @@ def transform(self, X):
+        X, Xt, mask_missing_values = self._initial_imputation(X)
 
-        X_filled = np.clip(X_filled, self._min_value, self._max_value)
+        # edge case: in case the user specifies 0 for n_iter,
```
I would remove this comment

glemaitre commented on sklearn/impute.py:
```diff
@@ -835,66 +867,68 @@ def fit_transform(self, X, y=None):
+        X, Xt, mask_missing_values = self._initial_imputation(X)
 
-        X_filled = np.clip(X_filled, self._min_value, self._max_value)
+        # edge case: in case the user specifies 0 for n_iter,
```
I would remove this comment

glemaitre commented on sklearn/impute.py:
```diff
@@ -835,66 +867,68 @@ def fit_transform(self, X, y=None):
+        # edge case: in case the user specifies 0 for n_iter,
+        # then there is no need to do further imputation and the result should
+        # be just the initial imputation (before clipping)
+        if self.n_iter < 1:
```
I would even restrict to

```python
if self.n_iter == 0
```

glemaitre replied:
Thought we need to check that it is > 0 at fit time

glemaitre commented on sklearn/impute.py:
```diff
@@ -916,44 +950,36 @@ def transform(self, X):
+        # edge case: in case the user specifies 0 for n_iter,
+        # then there is no need to do further imputation and the result should
+        # be just the initial imputation (before clipping)
+        if self.n_iter < 1:
```
I would even restrict to

```python
if self.n_iter == 0
```

glemaitre reviewed (commented):
It is missing a test when n_iter = 0 which should give the exact results of the initial imputer.

jorisvandenbossche commented on doc/modules/impute.rst:
```diff
@@ -128,17 +128,37 @@ Multiple vs. Single Imputation
+
+Our implementation of :class:`IterativeImputer` was inspired by the R MICE
+package (Multivariate Imputation by Chained Equations) [1], but differs from
+it in setting single imputation to default instead of multiple imputation.
```
We can update the text here in a similar way that the same text in the docstring was updated (about the "setting single imputation to default" which is not fully correct)

jorisvandenbossche commented on examples/plot_missing_values.py:
```diff
@@ -95,7 +95,7 @@ def get_results(dataset):
             'Zero imputation',
             'Mean Imputation',
-            'Chained Imputation']
+            'Iterative Imputation']
```
maybe also add "multivariate", or does the label get too long then?

jorisvandenbossche commented on sklearn/impute.py:
```diff
@@ -499,31 +511,42 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+    This implementation was inspired by the R MICE package (Multivariate
+    Imputation by Chained Equations), but differs from it by returning a single
+    imputation instead of multiple imputations. However, multiple imputation is
+    supported with multiple instances of the imputer with different random
```
"is supported" -> "can be achieved" ? (for me that is clearer that it is still some work to do)

jorisvandenbossche commented on sklearn/impute.py:
```diff
@@ -499,31 +511,42 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+    supported with multiple instances of the imputer with different random
+    seeds run in parallel.
+
+    To support imputation in inductive mode we store each feature's predictor
```
"inductive mode" is not clear for me. This means "during transform" ? 
Maybe we can only state that the predictors are stored so they can be used to predict without refitting during transform?

jorisvandenbossche commented:
> I don't think the idea of convergence makes sense when you're doing multiple imputation. I'm okay leaving it as n_iter.

I still think ideally we check convergence in the future (and thus a `max_iter` would be more appropriate). 
We can always say that in case of `sample_posterior=True`, no early stopping is used?

sergeyf commented:
@GaelVaroquaux @glemaitre Re: setting RF to be default. This means a *much* slower imputer. Why does it matter that having `RidgeCV` makes `IterativeImputer` more like matrix factorization? I think this is actually an interesting point that warrants an example comparing the two in speed and accuracy, but not something that warrants changing the default for `IterativeImputer` from the most practical estimator.

GaelVaroquaux commented:
> @GaelVaroquaux @glemaitre Re: setting RF to be default. This means a much
> slower imputer.

Yes. But MICE / Chain imputer is not know as a fast method, rather one
for powerful models.

> Why does it matter that having RidgeCV makes IterativeImputer more like
> matrix factorization? I think this is actually an interesting point
> that warrants an example comparing the two in speed and accuracy, but
> not something that warrants changing the default for IterativeImputer
> from the most practical estimator.

I don't understand: if you care about speed, you should be using low-rank
imputation, which will be faster than MICE with linear models.

glemaitre committed:
iter

glemaitre committed:
Merge remote-tracking branch 'origin/master' into chained_imputer_update

GaelVaroquaux commented:
I suspect that you are using a linear model as a supervised learner to do
the benchmarks?

In missing data imputation for analysis, there is a condition that the
imputation model performs best if it is "congenial" with the analytic
model, in other words that the procedure for the analysis can be derived
from the model generating the imputation
(https://www.jstor.org/stable/pdf/2246252.pdf).

We should probably add this in the docs, explaining how to choose the
model for the imputer (simple rule: if the analytic model is simple, use
a simple model, if it's complex, use a complex model).

glemaitre commented:
> I suspect that you are using a linear model as a supervised learner to do the benchmarks?

I was using a RandomForestRegressor.

GaelVaroquaux commented:
> I was using a RandomForestRegressor.

Ah, ah, ah.

Well, I'll take that. So much for my pedantic theory (though technically,
it's not wrong, because RandomForest is congenial wrt linear models,
while the converse is not true).

glemaitre commented:
```python
import numpy as np
import pandas as pd

from sklearn.datasets import load_boston
from sklearn.datasets import load_diabetes
from sklearn.impute import IterativeImputer
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import ParameterGrid

rng = np.random.RandomState(42)

X, y = load_diabetes(return_X_y=True)

n_missing_samples = X.shape[0] // 2
n_missing_features = X.shape[1] // 2

for feature_idx in rng.choice(np.arange(X.shape[1]), size=n_missing_features, replace=False):
    samples_idx = rng.choice(np.arange(X.shape[0]), size=n_missing_samples, replace=False)
    X[samples_idx, feature_idx] = np.nan

pipe = make_pipeline(IterativeImputer(random_state=0),
                     RandomForestRegressor(random_state=0))
grid = GridSearchCV(pipe, {'iterativeimputer__n_iter': [10, 5, 3, 1, 0],
                           'iterativeimputer__predictor': [None, RandomForestRegressor(n_estimators=100)]},
                   return_train_score=False)
grid.fit(X, y)
df = pd.DataFrame(grid.cv_results_)
df[['param_iterativeimputer__n_iter', 'param_iterativeimputer__predictor',
    'mean_test_score', 'std_test_score', 'rank_test_score']]
```

```
 param_iterativeimputer__n_iter  \
0                             10   
1                             10   
2                              5   
3                              5   
4                              3   
5                              3   
6                              1   
7                              1   
8                              0   
9                              0   

                   param_iterativeimputer__predictor  mean_test_score  \
0                                               None         0.373361   
1  RandomForestRegressor(bootstrap=True, criterio...         0.350114   
2                                               None         0.362885   
3  RandomForestRegressor(bootstrap=True, criterio...         0.348411   
4                                               None         0.376764   
5  RandomForestRegressor(bootstrap=True, criterio...         0.350230   
6                                               None         0.359150   
7  RandomForestRegressor(bootstrap=True, criterio...         0.325017   
8                                               None         0.283729   
9  RandomForestRegressor(bootstrap=True, criterio...         0.283729   

   std_test_score  rank_test_score  
0        0.045388                2  
1        0.028443                6  
2        0.041901                3  
3        0.062795                7  
4        0.038612                1  
5        0.037106                5  
6        0.062425                4  
7        0.051899                8  
8        0.095550                9  
9        0.095550                9  
```

glemaitre commented:
> Well, I'll take that. So much for my pedantic theory (though technically, it's not wrong, because RandomForest is congenial wrt linear models, while the converse is not true).

I would be careful because there is too much variability in the results depending of the generation of the missing data. Not sure how we could do that properly.

glemaitre commented:
Ok some thoughts from the sprint here:

The default values that we should set are not obvious. We should do some more extensive benchmark with different datasets and try to make sense of it. We should not rush it for the release, so therefore we propose to keep this for the next release (0.21), which means reverting what is in master for now.

sergeyf commented:
@glemaitre Just to clarify, what does this refer to: "which means reverting what is in master for now"? Will the PR be merged at all? A lot of the changes are documentation, bug fixes, and naming.

glemaitre commented:
> @glemaitre Just to clarify, what does this refer to: "which means reverting what is in master for now"? Will the PR be merged at all? A lot of the changes are documentation, bug fixes, and naming.

I think that we are going to remove totally the ChainedImputer from master to not have it in 0.20. We will re-introduce it in 0.21 and also bring this PR and set the default after looking at an extensive benchmark.

In short, if we make a mistake regarding naming or default parameter, it will be then really costly since that we will have a 2 versions of deprecation or future warning which is a pain for users.

sergeyf commented:
Oh bummer. That means another year's delay, probably?

glemaitre commented:
> Oh bummer. That means another year's delay, probably?

I hope that we will release sooner.

sergeyf commented:
What kind of help can I provide in the meanwhile?

jnothman commented:
Delaying this also makes me a bit sad, but users who want it will fetch master, hopefully in a few weeks, when we have ironed things out. I agree that we should release it with reasonable defaults.

I also feel responsible for this, because in reviewing, I did not pursue my intuition that this should be more general than to require sampling from posteriors, nor did I properly look into MICE...

Should we throw it into a branch, as we did with discrete, so that we can work on this incrementally (including @RianneSchouten's multiple imputation example)?

glemaitre commented:
> Should we throw it into a branch, as we did with discrete, so that we can work on this incrementally 

I think that this is the right thing to do

GaelVaroquaux commented:
I feel the same. Imputation is something that we (scikit-learn team) are
only starting to learn. We don't know the good practices and the sane
defaults.

We should try to release more often than once a year, to make postponing
features like this less painful.

@sergeyf: what you can do to help, is to help us build examples that
convey the compromises and help users (and library developers) to make
the right choices.

GaelVaroquaux commented:
> Should we throw it into a branch, as we did with discrete, so that we can work on this incrementally (including @RianneSchouten's multiple imputation example)?

+1 (including the example)

sergeyf commented:
@GaelVaroquaux Gotcha, thanks. I'll ponder some useful examples when I have time next.

jorisvandenbossche commented:
OK, I started with the first step, reverting in master: https://github.com/scikit-learn/scikit-learn/pull/11600

After that (when that is merged), I will create a new branch adding that stuff again, and then this PR and the PR with the example can be updated to target that branch (that can be done without closing the PR).

jorisvandenbossche commented:
OK, the revert (https://github.com/scikit-learn/scikit-learn/pull/11600) is merged. Next, I will make a branch that adds it back, and rebase this PR on top of that (but will first focus on release critical things)

jimmywan commented on examples/plot_missing_values.py:
```diff
@@ -4,11 +4,11 @@
 which could dominate results (otherwise known as a 'long tail').
 
-Another option is the :func:`sklearn.impute.ChainedImputer`. This uses
+Another option is the :class:`sklearn.impute.ChainedImputer`. This uses
```
Looks like you missed renaming `ChainedImputer` to `IterativeImputer` here.

sergeyf replied:
Ah, thanks.

sergeyf committed:
Missing rename

jnothman commented:
I look forward to getting this right immediately after the 0.20 release!

sergeyf committed:
Merge branch 'iterativeimputer' into chained_imputer_update

sergeyf commented:
@jnothman This was a strange merge conflict. It looked like both of the conflicting code snippets were something to add, so I did that. Please take a look and see if you agree. What are next steps here?

sergeyf committed:
Undoing changes to 0.20 rst

Impute not part of 0.20!

sergeyf committed:
Updating 0.21 changes to IterativeImputer

sergeyf committed:
Update v0.20 rst to not have any changes

jnothman commented:
I think you did an unfinished rename of a variable:

flake8 failures
```
examples/plot_missing_values.py:81:5: F841 local variable 'chained_impute_scores' is assigned to but never used
    chained_impute_scores = cross_val_score(estimator, X_missing, y_missing,
    ^
examples/plot_missing_values.py:87:14: F821 undefined name 'iterative_impute_scores'
            (iterative_impute_scores.mean(), iterative_impute_scores.std()))
             ^
examples/plot_missing_values.py:87:46: F821 undefined name 'iterative_impute_scores'
            (iterative_impute_scores.mean(), iterative_impute_scores.std()))
                                             ^
```

Error in examples:
```

Unexpected failing examples:
/home/circleci/project/examples/plot_missing_values.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/circleci/project/examples/plot_missing_values.py", line 90, in <module>
    results_diabetes = np.array(get_results(load_diabetes()))
  File "/home/circleci/project/examples/plot_missing_values.py", line 87, in get_results
    (iterative_impute_scores.mean(), iterative_impute_scores.std()))
NameError: name 'iterative_impute_scores' is not defined
```

sergeyf committed:
Fix plot_missing_values.py

sergeyf commented:
Thanks @jnothman. This is kind of weird: I remember all tests passing just fine before all the unmerging started happening.

sergeyf commented:
@glemaitre it still says "requested changes" from you. Do you remember what these were?

sergeyf commented:
Great, thanks.

On Tue, Sep 4, 2018, 7:23 PM Guillaume Lemaitre <notifications@github.com>
wrote:

> I would need to review once more since that a lot of changes have been
> done. For sure my previous comments have been addressed in some way.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11350#issuecomment-418487442>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABya7BBNbcaQRCLXAqyWDsaNJev4o45qks5uXtMtgaJpZM4U1b8P>
> .
>

jnothman reviewed (approved):
Minor comments. I'd like to merge this soon, but if you could please pull together a list of remaining questions/issues/goals, @sergeyf, that would be great

sergeyf commented on sklearn/impute.py:
```diff
@@ -487,31 +499,42 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
-    additional fitting. We do this by storing each feature's predictor during
-    the round-robin ``fit`` phase, and predicting without refitting (in order)
-    during the ``transform`` phase.
+    This implementation was inspired by the R MICE package (Multivariate
```
Done.

sergeyf commented on sklearn/impute.py:
```diff
@@ -449,24 +452,33 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        If ``sample_posterior`` is True, the predictor must support
+        ``return_std`` in its ``predict`` method. Also, if
+        ``sample_posterior=True`` the default predictor will be
+        ``BayesianRidge()`` and ``RidgeCV`` otherwise.
```
Done.

sergeyf committed:
Addressing review comments

sergeyf commented:
@jnothman thanks for the review.

In terms of outstanding issues, I'm not that sure actually. 

I think there was a concern that `n_iter` should be called `max_iter` if we ever use early stopping.

We also still need a solid multiple imputation example.

Gael asked for "examples that convey the compromises and help users (and library developers) to make the right choices." I think he meant "what are sane defaults?"

We probably need to do a survey of the packages in R, figure out which ones have been battle tested and are well-respected, and see if `IterativeImputer` can do what they do as it currently exists.

Am I missing something obvious?

sergeyf commented:
@jnothman lots of weird test errors: `SyntaxError: invalid escape sequence \_`  I am almost sure this isn't my fault!

jnothman commented on sklearn/impute.py:
```diff
@@ -449,24 +452,33 @@ class ChainedImputer(BaseEstimator, TransformerMixin):
+        Whether to sample from the (Gaussian) predictive posterior of the
+        fitted predictor for each imputation. Predictor must support
+        ``return_std`` in its ``predict`` method if set to ``True``. Set to
+        ``True`` if using ``IterativeImputer`` for multiple imputations.
```
I'm not sure how to pick one of the trees uniformly at random in a sensible way. But the API could be extended to support different ways of sampling the posterior via a non-binary switch here, i.e. sample_posterior='std' or something

sergeyf replied:
Ok I can investigate if we agree it actually makes sense. I am not even sure how to tell if it does.

jnothman commented:
I've merged in master where that CI issue was fixed.

jnothman commented:
Yes, I think you're right to make further changes driven by examples: we want an example that illustrates something along the lines of missforest functionality, and another along the lines of MICE, and use that to identify appropriate. default behaviour.

jnothman commented on sklearn/impute.py:
```diff
@@ -462,7 +462,8 @@ class IterativeImputer(BaseEstimator, TransformerMixin):
         ``return_std`` in its ``predict`` method. Also, if
         ``sample_posterior=True`` the default predictor will be
-        ``BayesianRidge()`` and ``RidgeCV`` otherwise.
+        ``:class:sklearn.linear_model:BayesianRidge()`` and
```
This is not the right syntax. Git grep :class: for examples

sergeyf committed:
Fixed :class: syntax

sergeyf commented:
@jnothman I'm concerned that a few examples won't tell us about sensible defaults. What we actually need is a large-enough set of experiments where we sweep meaningful parameters to get a sense of reasonable defaults. What's big enough? How has `sklearn` previous decided on defaults if it's unclear?

jnothman commented:
Here we know less about datasets than about existing implementations. But
we could use Rianne's amputation to help apply this to assorted
classification or regression tasks. But there are a lot of variables in the
problem.

I actually think defaults are less important here than making sure it's
easy to get commonly used functionality, like what you would expect of
missforest.

One other thing that may be of interest is the ability to impute
categorical variables. I.e. have a categorical_features mask and accept a
classifier as well as regressor for imputation. But that can certainly come
with time.

sergeyf commented:
Ok makes sense. All tests pass now. I'll wait for this to merge and then build a missForest example on top of what we currently have. Let me know if there's anything else for this PR.

jnothman commented:
Thanks!
